pkg/kubelet/logs/container_log_manager.go

```diff
@@ -0,0 +1,401 @@
+	MaxFiles int
+}
+
+// GetAllLogs gets all inuse (rotated/compressed) logs for a specific container log.
+// Returned logs are sorted in oldest to newest order.
+// TODO(#59902): Leverage this function to support log rotation in `kubectl logs`.
+func GetAllLogs(log string) ([]string, error) {
+	// pattern is used to match all rotated files.
+	pattern := fmt.Sprintf("%s.*", log)
+	logs, err := filepath.Glob(pattern)
+	if err != nil {
+		return nil, fmt.Errorf("failed to list all log files with pattern %q: %v", pattern, err)
+	}
+	inuse, _ := filterUnusedLogs(logs)
+	sort.Strings(inuse)
+	return append(inuse, log), nil
+}
+
+// compressReadCloser wraps gzip.Reader with a function to close file handler.
+type compressReadCloser struct {
+	f *os.File
+	*gzip.Reader
+}
+
+func (rc *compressReadCloser) Close() error {
+	ferr := rc.f.Close()
+	rerr := rc.Reader.Close()
+	if ferr != nil {
+		return ferr
+	}
+	if rerr != nil {
+		return rerr
+	}
+	return nil
+}
+
+// UncompressLog compresses a compressed log and return a readcloser for the
+// stream of the uncompressed content.
+// TODO(#59902): Leverage this function to support log rotation in `kubectl logs`.
+func UncompressLog(log string) (_ io.ReadCloser, err error) {
+	if !strings.HasSuffix(log, compressSuffix) {
+		return nil, fmt.Errorf("log is not compressed")
+	}
+	f, err := os.Open(log)
+	if err != nil {
+		return nil, fmt.Errorf("failed to open log: %v", err)
+	}
+	defer func() {
+		if err != nil {
+			f.Close()
+		}
+	}()
+	r, err := gzip.NewReader(f)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create gzip reader: %v", err)
+	}
+	return &compressReadCloser{f: f, Reader: r}, nil
+}
+
+// parseMaxSize parses quantity string to int64 max size in bytes.
+func parseMaxSize(size string) (int64, error) {
+	quantity, err := resource.ParseQuantity(size)
+	if err != nil {
+		return 0, err
+	}
+	maxSize, ok := quantity.AsInt64()
+	if !ok {
+		return 0, fmt.Errorf("invalid max log size")
+	}
+	return maxSize, nil
+}
+
+type containerLogManager struct {
+	runtimeService internalapi.RuntimeService
+	policy         LogRotatePolicy
+	clock          clock.Clock
+}
+
+// NewContainerLogManager creates a new container log manager.
+func NewContainerLogManager(runtimeService internalapi.RuntimeService, maxSize string, maxFiles int) (ContainerLogManager, error) {
+	if maxFiles <= 0 {
+		return nil, fmt.Errorf("invalid MaxFiles %d, must be > 0", maxFiles)
+	}
+	parsedMaxSize, err := parseMaxSize(maxSize)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse container log max size %q: %v", maxSize, err)
+	}
+	// policy LogRotatePolicy
+	return &containerLogManager{
+		runtimeService: runtimeService,
+		policy: LogRotatePolicy{
+			MaxSize:  parsedMaxSize,
+			MaxFiles: maxFiles,
+		},
+		clock: clock.RealClock{},
+	}, nil
+}
+
+// Start the container log manager.
+func (c *containerLogManager) Start() {
+	if c.policy.MaxSize < 0 {
```

Comment:
can we instead perform validation when we parse MaxSize it and return an error there?  We generally fail the kubelet when given bad config.

Reply:
Or is this intentional?

Reply:
If this is meant to disable it, can we instead just use the feature gate for disabling this?  Is there a use-case we know of where we will not want the kubelet to rotate logs.